src/app/loadout-drawer/loadout-apply.ts

```diff
@@ -798,73 +976,90 @@ function applyLoadoutMods(
-    return successfulMods;
-  };
+  // Early exit - if all the mods are already there, nothing to do
+  return modHashes.every((h) => {
+    const foundAt = existingMods.indexOf(h);
+    if (foundAt === -1) {
+      // a mod was missing
+      return false;
+    } else {
+      // the mod was found, but we have consumed this copy of it
+      delete existingMods[foundAt];
+      return true;
+    }
+  });
 }
 
 /**
- * Equip the specified mods on the item, in the order provided.
- * This applies each assignment, and does not account item energy,
- * which should be pre-calculated .
+ * Equip the specified mods on the item, in the order provided. This applies
+ * each assignment, and does not account for item energy, which should be
+ * pre-calculated.
  */
 function equipModsToItem(
   itemId: string,
   modsForItem: Assignment[],
+  /** Callback for state reporting while applying. Mods are applied in parallel so we want to report ASAP. */
+  onSuccess: (assignment: Assignment) => void,
+  /** Callback for state reporting while applying. Mods are applied in parallel so we want to report ASAP. */
+  onFailure: (assignment: Assignment, error?: Error, equipNotPossible?: boolean) => void,
   includeAssignToDefault = false
-): ThunkResult<{ successfulMods: number[]; errors: Error[] }> {
+): ThunkResult {
   return async (dispatch, getState) => {
     const defs = d2ManifestSelector(getState())!;
     const item = getItemAcrossStores(storesSelector(getState()), { id: itemId })!;
     const destiny2CoreSettings = destiny2CoreSettingsSelector(getState())!;
 
     if (!item.sockets) {
-      return { successfulMods: emptyArray(), errors: emptyArray() };
+      return;
     }
 
     const modsToApply = [...modsForItem];
-    const successfulMods: number[] = [];
-    const errors: Error[] = [];
 
     // TODO: we tried to do these applies in parallel, but you can get into trouble
     // if you need to remove a mod before applying another.
 
-    for (const { socketIndex, mod } of modsToApply) {
+    for (const assignment of modsToApply) {
+      const { socketIndex, mod } = assignment;
       // Use this socket
       const socket = getSocketByIndex(item.sockets, socketIndex)!;
       // If the plug is already inserted we can skip this
       if (socket.plugged?.plugDef.hash === mod.hash) {
         // Don't count removing mods as applying a mod successfully
-        if (includeAssignToDefault || !isAssigningToDefault(item, { socketIndex, mod }, defs)) {
-          successfulMods.push(mod.hash);
+        if (includeAssignToDefault || !isAssigningToDefault(item, assignment, defs)) {
+          onSuccess(assignment);
         }
         continue;
       }
-      const defaultPlugHash = getDefaultPlugHash(socket, defs);
-      if (defaultPlugHash && canInsertPlug(socket, defaultPlugHash, destiny2CoreSettings, defs)) {
+      if (canInsertPlug(socket, mod.hash, destiny2CoreSettings, defs)) {
```

Comment:
Pretty sure this was a bug? I'm not sure why we were checking whether we can insert the default plug instead of the actual plug.